app/controllers/scim_v2/users_controller.rb

```diff
@@ -0,0 +1,103 @@
+    skip_before_action :verify_authenticity_token
+
+    rescue_from "ActiveRecord::RecordNotFound", with: :handle_resource_not_found
+
+    def index
+      query = if params[:filter].blank?
+                storage_scope
+              else
+                attribute_map = storage_class.new.scim_queryable_attributes
+                parser        = ::Scimitar::Lists::QueryParser.new(attribute_map)
+
+                parser.parse(params[:filter])
+                parser.to_activerecord_query(storage_scope)
+              end
+
+      pagination_info = scim_pagination_info(query.count)
+      page_of_results = query
+                          .order(id: :asc)
+                          .offset(pagination_info.offset)
+                          .limit(pagination_info.limit)
+                          .to_a
+
+      super(pagination_info, page_of_results) do |record|
+        record.to_scim(location: url_for(action: :show, id: record.id))
+      end
+    end
+
+    def show
+      super do |user_id|
+        user = storage_scope.find(user_id)
+        user.to_scim(location: url_for(action: :show, id: user_id))
+      end
+    end
+
+    def create
+      super do |scim_resource|
+        storage_class.transaction do
+          user = storage_class.new
+          user.from_scim!(scim_hash: scim_resource.as_json)
+          call = Users::CreateService
+                   .new(user: User.system)
+                   .call(user.attributes)
+                   .on_failure { |result| raise result.message }
+
+          user = call.result
+          user.to_scim(location: url_for(action: :show, id: user.id))
+        end
+      end
+    end
+
+    def replace
+      super do |user_id, scim_resource|
+        storage_class.transaction do
+          user = storage_scope.find(user_id)
+          user.from_scim!(scim_hash: scim_resource.as_json)
+          Users::UpdateService
+            .new(user: User.system, model: user)
+            .call
+            .on_failure { |call| raise call.message }
+          user.to_scim(location: url_for(action: :show, id: user.id))
+        end
+      end
+    end
+
+    def update
+      super do |user_id, patch_hash|
+        storage_class.transaction do
+          user = storage_scope.find(user_id)
+          user.from_scim_patch!(patch_hash: patch_hash)
+          Users::UpdateService
+            .new(user: User.system, model: user)
+            .call
+            .on_failure { |call| raise call.message }
+          user.to_scim(location: url_for(action: :show, id: user.id))
+        end
+      end
+    end
+
+    def destroy
+      super do |user_id|
+        user = storage_scope.find(user_id)
+        Users::DeleteService
+          .new(user: User.system, model: user)
+          .call
+          .on_failure { |call| raise call.message }
+      end
+    end
+
+    protected
+
+    def storage_class
+      User
+    end
+
+    def storage_scope
+      User.user.where.not(identity_url: [nil, ""])
```

Comment:
I am pretty sure that this can be simplified to

```suggestion
      User.where.not(identity_url: [nil, ""])
```

For comparision:

```ruby
[2] pry(main)> User.all.to_sql
=> "SELECT \"users\".* FROM \"users\" WHERE \"users\".\"type\" = 'User'"
[3] pry(main)> User.all.user.to_sql
=> "SELECT \"users\".* FROM \"users\" WHERE \"users\".\"type\" = 'User'"
[4] pry(main)> Principal.all.user.to_sql
=> "SELECT \"users\".* FROM \"users\" WHERE \"users\".\"type\" = 'User'"
[5] pry(main)> Principal.all.to_sql
=> "SELECT \"users\".* FROM \"users\""
```

Reply:
An unrelated thought (GitHub please, give me multiple threads per line of code!):

:red_circle: I am not sure that we should completely hide users from view that don't have an identity url. I agree that we can't match them through their external id, but now the SCIM client also can't match them through other factors such as an email address. Thus conflicts between existing "local" users and newly created "SCIM" users are unavoidable (e.g. duplicate email addresses in our validations).

That said, I am not sure whether any SCIM client is attempting to avoid such conflicts, but I would definitely check on the scim documentation on whether they say anything about returning users without an external id (e.g. whether we are _required_ to hide those users).

:red_circle: On the other hand we should probably try to filter out `builtin?` users. Though as far as I can tell, they are not database filterable... so we will have to come up with something there.

Reply:
> 🔴 I am not sure that we should completely hide users from view that don't have an identity url. I agree that we can't match them through their external id, but now the SCIM client also can't match them through other factors such as an email address. Thus conflicts between existing "local" users and newly created "SCIM" users are unavoidable (e.g. duplicate email addresses in our validations).
> 
> That said, I am not sure whether any SCIM client is attempting to avoid such conflicts, but I would definitely check on the scim documentation on whether they say anything about returning users without an external id (e.g. whether we are _required_ to hide those users).

Accroding to specs(https://www.rfc-editor.org/rfc/rfc7643.html#section-3.1) `externalId` is optional. So, the scope should not have **identity_url** constraint.

> 🔴 On the other hand we should probably try to filter out `builtin?` users. Though as far as I can tell, they are not database filterable... so we will have to come up with something there.

There is `not_builtin` scope. I will use it.